### ISSUE TITLE
[@types/sinon] massive changes between 5.0.1 and 5.0.7

### ISSUE THREAD

**Issue by bradtaniguchi** [DefinitelyTyped/DefinitelyTyped#30702]
If you know how to fix the issue, make a pull request instead.

- [x] I tried using the `@types/sinon#5.0.7` package and had problems. (5.0.7 is latest)
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [X] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond. (sorry for including everyone!)
  - Authors: @mrbigdog2u, @rationull, @lumaxis, @nicojs, @43081j, @joshuakgoldberg, @gjednaszewski, @johnjesse

----
I recently installed the latest @types/sinon package (version 5.0.7) and found a number of **breaking changes** compared to the 5.0.1 package. Resulting in a number of spec file errors. These sorts of errors were very unexpected due the version changing being only a "patch", but resulted in 100+ linting errors.

Most of these seem to be related to the changes made to the sinon.SinonSandbox type.  For example, this code used to be fine in 5.0.1
```typescript
import * as sinon from 'sinon';
// describe
let sandbox: sinon.SinonSandbox;
beforeEach(() => {
  sandbox = sinon.createSanbox();
});
// spec file
sandbox
  .stub(AppModel.prototype as any, 'create')
  .returns(Promise.resolve({}));
// class file
AppModel {
  private create(): AppResponse  {
    //...
  }
}
```
But after 5.0.7 installation I'd get errors on the given code such as:
```
test/models/app-model.spec.ts(59,7): error TS2349: Cannot invoke an expression 
whose type lacks a call signature. 
Type '((obj: any) => SinonStub<any[], any>) | ((obj: {}) => SinonStub<unknown[], {}>)' 
has no compatible call signatures.
```
I was able to "quick-fix" this issue by changing the spec file code to:
```typescript
(sandbox as any)
  .stub(AppModel.prototype as any, 'create')
  .returns(Promise.resolve({}));
```
But this sort of solution was not optimal since I was essentially making sinon's sandbox an any type just to fix issues with my "stubbing". This issue really caught me off guard since it was introduced within a "patch" change. 

There were a number of other unwanted changes that all resulted in a number of linting errors.

**Comment by 43081j** [comment]
pretty much a duplicate of #30657.

waiting on DT maintainers to advise.

**Comment by amitdahan** [comment]
@43081j why is this a duplicate? Different errors.

And just to clarify - this is caused by misinference of TArgs and TReturnType of the `.stub` when the first arg is an `any`.
Thus the breakage of future calls to function-prototype-restricting function calls like `callsFake` or most behavior changing functions.

The more strict type checks are welcome in my eyes, except the fact that `any` is being treated in a weird way.

**Comment by 43081j** [comment]
Because all these errors being brought up in the past week are caused by the same release.

A bad merge to master is the source of all our problems. I'm trying to fix that in my #30856 PR.

It should have been a breaking change version bump and the type parameters are a little too strict.

**Comment by fatso83** [comment]
This should be closed

**Comment by orta** [comment]
Hi thread, we're moving DefinitelyTyped to use [GitHub Discussions](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/53377) for conversations the `@types` modules in DefinitelyTyped.

To help with the transition, we're closing all issues which haven't had activity in the last 6 months, which includes this issue. If you think closing this issue is a mistake, please pop into the [TypeScript Community Discord](https://discord.gg/typescript) and mention the issue in the `definitely-typed` channel.
